### PR TITLE
[Core] Add support for inferring Annotated types

### PR DESCRIPTION
**Pull request by hinthornw**
in bind_tools() / convert_to_openai_function

Still really low suport for other things... still think we should just re-use the tool decorator

### CODE REVIEW

File: libs/core/langchain_core/utils/function_calling.py [hunk 1]
```diff
@@ -171,10 +176,27 @@ def _get_python_function_arguments(function: Callable, arg_descriptions: dict) -
...
-            # "type" has no attribute "schema"
-            properties[arg] = arg_type.schema()  # type: ignore[attr-defined]
+
+        if _is_annotated_type(arg_type):
+            annotated_args = get_args(arg_type)
+            arg_type = annotated_args[0]
+            if len(annotated_args) > 1:
+                for annotation in annotated_args[1:]:
+                    if isinstance(annotation, str):
+                        arg_descriptions[arg] = annotation
```
Comment (hinthornw) [inline libs/core/langchain_core/utils/function_calling.py:186]: at some point could add other fields like "examples", etc if we wanted

File: libs/core/langchain_core/utils/function_calling.py [hunk 2]
```diff
@@ -185,13 +207,20 @@ def _get_python_function_arguments(function: Callable, arg_descriptions: dict) -
             and getattr(arg_type, "__dict__").get("__origin__", None) == Literal
         ):
             properties[arg] = {
-                "enum": list(arg_type.__args__),  # type: ignore
-                "type": PYTHON_TO_JSON_TYPES[arg_type.__args__[0].__class__.__name__],  # type: ignore
+                "enum": list(arg_type.__args__),
+                "type": PYTHON_TO_JSON_TYPES[arg_type.__args__[0].__class__.__name__],
             }
+        else:
+            logger.warning(
```
Comment (hinthornw) [inline libs/core/langchain_core/utils/function_calling.py:214]: Previously we were silently just not including args in the schema, which is pretty annoying

File: libs/core/langchain_core/utils/function_calling.py [hunk 3]
```diff
@@ -171,10 +176,27 @@ def _get_python_function_arguments(function: Callable, arg_descriptions: dict) -
...
+                for annotation in annotated_args[1:]:
+                    if isinstance(annotation, str):
+                        arg_descriptions[arg] = annotation
+                        break
+        if (
+            isinstance(arg_type, type)
+            and hasattr(arg_type, "model_json_schema")
+            and callable(arg_type.model_json_schema)
+        ):
+            properties[arg] = arg_type.model_json_schema()
```
Comment (baskaryan) [inline libs/core/langchain_core/utils/function_calling.py:193]: this is for pydantic v2 support?

Comment (hinthornw) [reply libs/core/langchain_core/utils/function_calling.py:193]: ya. I mean both work, but this doesn't give a warning

**Pull request merged**